Storage/tests/Unit/Connection/RestTest.php

```diff
@@ -469,6 +471,208 @@ public function validationMethod()
+        // private method's logic by verifying the output for a given input.
+        $rest = new Rest($restConfig);
+        $reflector = new \ReflectionClass('Google\Cloud\Storage\Connection\Rest');
+        $method = $reflector->getMethod('isPreConditionSupplied');
+        $condIdempotentOps = $reflector->getProperty('condIdempotentOps');
+        $method->setAccessible(true);
+        $condIdempotentOps->setAccessible(true);
+        $condIdempotentOps = $condIdempotentOps->getValue($rest);
+        $methodName = sprintf('%s.%s', $resource, $op);
+        $preconditionNeeded = array_key_exists($methodName, $condIdempotentOps);
+        $result = $method->invokeArgs($rest, array($methodName, $args));
+        if ($preconditionNeeded and $errorCode != 400) {
+            $this->assertEquals($result, $expected);
+        } else {
+            // This case will imply:
+            //     * Precondition needed & errorCode == 400 => No retry
+            //     * Precondition non needed => idempotent / non idempotent op
+            // In both the case, this function's response will never be taken
+            // into consideration for deciding whether to retry or not. Thus
+            // we can just put a dummy assertion here.
+            $this->assertTrue(true);
+        }
+    }
+
+    /**
+     * @dataProvider retryFunctionReturnValues
+     * @dataProvider retryStrategyCases
+     */
+    public function testRetryFunction(
+        $resource,
+        $op,
+        $restConfig,
+        $args,
+        $errorCode,
+        $currAttempt,
+        $expected
+    ) {
+        $rest = new Rest($restConfig);
+        $reflection = new \ReflectionClass($rest);
+        $property = $reflection->getProperty('restRetryFunction');
+        $property->setAccessible(true);
+        $restRetryFunction = $property->getValue($rest);
+        $retryFun = $rest->getRestRetryFunction($resource, $op, $args, $restRetryFunction);
+
+        $this->assertEquals(
+            $expected,
+            $retryFun(new \Exception('', $errorCode), $currAttempt)
+        );
+    }
+
+    /**
+     * This tests whether the $arguments passed to the callbacks for header
+     * updation is properly done when those callbacks are invoked in the
+     * ExponentialBackoff::execute() method.
+     */
+    public function testUpdateRetryHeaders()
+    {
+        $arguments = [new Request('GET', '/somewhere'), ['headers' => []]];
+        $requestHash = 'dummy_hash';
+        $currentAttempt = 1;
+        $headerLineName = AgentHeader::AGENT_HEADER_KEY;
+
+        // Using Reflection instead of Prophecy because we want to test a
+        // private method's logic by verifying the output for a given input.
+        $rest = new Rest();
+        $reflection = new \ReflectionClass(Rest::class);
+        $reflectionMethod = $reflection->getMethod('updateRetryheaders');
+        $reflectionMethod->setAccessible(true);
```

Comment:
So I've been running these tests, and when I remove the implementation code in `Storage/Connection/Rest`, all the tests still pass. This is an issue because, even though we have tests for the `updateRetryHeaders` method, we don't have tests to ensure these headers are being added to the request.

This is why generally speaking, tests shouldn't be written to test private methods directly, but rather to test the functionality they provide from the public surface. 

So we need tests that mock the request object and call `Storage/Connection/Rest::send` and confirm the headers were added as expected.